exec/java-exec/src/main/java/org/apache/drill/exec/work/user/PlanSplitter.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.drill.exec.util.Utilities;
+import org.apache.drill.exec.work.QueryWorkUnit;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Helper class to return PlanFragments based on the query plan
+ * or based on split query plan
+ *
+ */
+public class PlanSplitter {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(PlanSplitter.class);
+
+  private static PlanSplitter s_instance = new PlanSplitter();
+
+  private PlanSplitter() {
+
+  }
+
+  public static PlanSplitter getInstance() {
+    return s_instance;
+  }
+
+  /**
+   * Method to plan the query and return list of fragments
+   * it will return query plan "as is" or split plans based on the req setting: split_plan
+   * @param dContext
+   * @param queryId
+   * @param req
+   * @param connection
+   * @return
+   */
+  public QueryPlanFragments planFragments(DrillbitContext dContext, QueryId queryId,
+      GetQueryPlanFragments req, UserClientConnection connection) {
+    QueryPlanFragments.Builder responseBuilder = QueryPlanFragments.newBuilder();
+    QueryContext queryContext = new QueryContext(connection.getSession(), dContext, queryId);
+
+    responseBuilder.setQueryId(queryId);
+
+    try {
+      responseBuilder.addAllFragments(getFragments(dContext, req, queryContext, queryId));
+      responseBuilder.setStatus(QueryState.COMPLETED);
+    } catch (Exception e) {
+      final String errorMessage = String.format("Failed to produce PlanFragments for query id \"%s\" with "
+          + "request to %s plan", queryId, (req.getSplitPlan() ? "split" : "no split"));
+      DrillPBError error = DrillPBError.newBuilder().setMessage(errorMessage).setErrorType(DrillPBError.ErrorType.PLAN).build();
+
+      responseBuilder.setStatus(QueryState.FAILED);
+      responseBuilder.setError(error);
+    }
+    return responseBuilder.build();
+  }
+
+  private List<PlanFragment> getFragments(final DrillbitContext dContext, final GetQueryPlanFragments req,
+      final QueryContext queryContext, final QueryId queryId) throws Exception {
+    final PhysicalPlan plan;
+    final String query = req.getQuery();
+    switch(req.getType()) {
+    case SQL:
+      final Pointer<String> textPlan = new Pointer<>();
+      plan = DrillSqlWorker.getPlan(queryContext, query, textPlan);
+      break;
+    case PHYSICAL:
+      plan = dContext.getPlanReader().readPhysicalPlan(query);
+      break;
+    default:
+      throw new IllegalStateException("Planning fragments supports only SQL or PHYSICAL QueryType");
+    }
+
+    MemoryAllocationUtilities.setupSortMemoryAllocations(plan, queryContext);
+
+    final PhysicalOperator rootOperator = plan.getSortedOperators(false).iterator().next();
+
+    final Fragment rootFragment = rootOperator.accept(MakeFragmentsVisitor.INSTANCE, null);
+    final SimpleParallelizer parallelizer = new SimpleParallelizerMultiPlans(queryContext);
+
+    List<PlanFragment> fragments = Lists.newArrayList();
+
+    if ( req.getSplitPlan() ) {
```

Comment:
Help me understand this please: at this point we for sure know that req is a split request. So why to make this check here again? 

If the intention is to allow the client to enable/disable splitting, why do we need this functionality?